eth_utils/logging.py

```diff
@@ -0,0 +1,89 @@
+from abc import ABCMeta
+import contextlib
+import logging
+from typing import Any, Type, Iterator
+
+from .toolz import assoc
+
+DEBUG2_LEVEL_NUM = 8
+
+
+class ExtendedDebugLogger(logging.Logger):
+    """
+    Logging class that can be used for lower level debug logging.
+    """
+
+    def debug2(self, message: str, *args: Any, **kwargs: Any) -> None:
+        self.log(DEBUG2_LEVEL_NUM, message, *args, **kwargs)
+
+
+def setup_DEBUG2_logging() -> None:
+    """
+    Installs the `DEBUG2` level logging levels to the main logging module.
+    """
+    if not hasattr(logging, "DEBUG2"):
+        logging.addLevelName(DEBUG2_LEVEL_NUM, "DEBUG2")
+        setattr(logging, "DEBUG2", DEBUG2_LEVEL_NUM)  # typing: ignore
+
+
+@contextlib.contextmanager
+def _use_logger_class(logger_cls: Type[logging.Logger]) -> Iterator:
+    original_logger_cls = logging.getLoggerClass()
+    logging.setLoggerClass(logger_cls)
+    try:
+        yield
+    finally:
+        logging.setLoggerClass(original_logger_cls)
+
+
+class HasLoggerMeta(ABCMeta):
+    """
+    Metaclass which using the `__qualname__` to assign a logger instance to the
+    class who's name corresponds to the import path for the class.
+    """
+
+    logger_class = logging.Logger
+
+    def __new__(mcls, name, bases, namespace):
+        if "logger" in namespace:
+            # If a logger was explicitely declared we shouldn't do anything to
+            # replace it.
+            return super().__new__(mcls, name, bases, namespace)
+        if "__qualname__" not in namespace:
+            raise AttributeError("Missing __qualname__")
```

Comment:
I've never really dived into meta classes. It all seems pretty magical to me still. Brian did another implementation that was also magical. Maybe he still has it floating around. I'm not sure which one was better. Also performance matters on some use cases, like py-evm/trinity. I don't know this code well enough to understand the performance impact.

Reply:
My understanding is that there should be no performance hit and this should more performant than the `@property` based approach since metaclass code runs at class creation time which is module loading time.  Assuming I am correct, there should be zero runtime overhead as everything should happen at module loading time.

Reply:
Just scrounged through my reflog to find my solution:

```python
class classproperty(Generic[TResult]):
    def __init__(self, func: Callable[[Type[Any]], TResult]) -> None:
        self.func = func

    def __get__(self, instance: Any, owner: Type[Any]) -> TResult:
        return self.func(owner)

def get_logger(cls: Type['BasePlugin']) -> logging.Logger:
    return logging.getLogger(f'trinity.extensibility.plugin#{cls.__name__}')

class BasePlugin(ABC):
    logger = classproperty(get_logger)
   ....
```

Where get_logger was pulled out to make mypy happy but could also be in-line with a `# type: ignore`

Never thought I'd say this but I expect that the metaclass is not only faster but also less magical.